File: csharp/SvyTests/Svy21Test.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using SVY21;

namespace SvyTests
{
    [TestClass]
    public class Svy21Test
    {
        private const double ToleranceLatLong = 0.0000000001;
        private const double ToleranceSvy21 = 0.001;

        [TestMethod]
        public void Svy21ToLatLongTest()
        {
            LatLongCoordinate expected, actual;
            Svy21Coordinate test;
            
            // SVY21 reference point.
            expected = new LatLongCoordinate(1.366666, 103.833333);
            test = new Svy21Coordinate(38744.572, 28001.642);
            actual = Svy21.ComputeLatitudeLongitude(test);
            Assert.AreEqual(expected.Latitude, actual.Latitude, ToleranceLatLong);
            Assert.AreEqual(expected.Longitude, actual.Longitude, ToleranceLatLong);

            // Corner of Ang Mo Kio 66kV Substation.
            expected = new LatLongCoordinate(1.3699278977737488, 103.856950349764668);
            test = new Svy21Coordinate(39105.269, 30629.967);
            actual = Svy21.ComputeLatitudeLongitude(test);
            Assert.AreEqual(expected.Latitude, actual.Latitude, ToleranceLatLong);
            Assert.AreEqual(expected.Longitude, actual.Longitude, ToleranceLatLong);
            
            // Corner of Jurong Lake Canal 400kV Cable Bridge.
            expected = new LatLongCoordinate(1.3446255443241177, 103.72794378041792);
            test = new Svy21Coordinate(36307.704, 16272.970);
            actual = Svy21.ComputeLatitudeLongitude(test);
            Assert.AreEqual(expected.Latitude, actual.Latitude, ToleranceLatLong);
            Assert.AreEqual(expected.Longitude, actual.Longitude, ToleranceLatLong);

            // Corner of Sembawang 66kV Substation.
            expected = new LatLongCoordinate(1.4520670518379692, 103.83080332777138);
            test = new Svy21Coordinate(48187.789, 27720.130);
            actual = Svy21.ComputeLatitudeLongitude(test);
            Assert.AreEqual(expected.Latitude, actual.Latitude, ToleranceLatLong);
            Assert.AreEqual(expected.Longitude, actual.Longitude, ToleranceLatLong);
        }

        [TestMethod]
        public void LatLongToSvy21Test()
        {
            Svy21Coordinate expected, actual;
            LatLongCoordinate test;

            // SVY21 Reference Point.
            expected = new Svy21Coordinate(38744.572, 28001.642);
            test = new LatLongCoordinate(1.366666, 103.833333);
            actual = Svy21.ComputeSvy21(test);
            Assert.AreEqual(expected.Northing, actual.Northing, ToleranceSvy21);
            Assert.AreEqual(expected.Easting, actual.Easting, ToleranceSvy21);

            // Corner of Ang Mo Kio 66kV Substation.
            expected = new Svy21Coordinate(39105.269, 30629.967);
            test = new LatLongCoordinate(1.3699278977737488, 103.85695034976466);
            actual = Svy21.ComputeSvy21(test);
            Assert.AreEqual(expected.Northing, actual.Northing, ToleranceSvy21);
            Assert.AreEqual(expected.Easting, actual.Easting, ToleranceSvy21);

            // Corner of Jurong Lake Canal 400kV Cable Bridge.
            expected = new Svy21Coordinate(36307.704, 16272.970);
            test = new LatLongCoordinate(1.3446255443241177, 103.72794378041792);
            actual = Svy21.ComputeSvy21(test);
            Assert.AreEqual(expected.Northing, actual.Northing, ToleranceSvy21);
            Assert.AreEqual(expected.Easting, actual.Easting, ToleranceSvy21);

            // Corner of Sembawang 66kV Substation.
            expected = new Svy21Coordinate(48187.789, 27720.130);
            test = new LatLongCoordinate(1.4520670518379692, 103.83080332777138);
            actual = Svy21.ComputeSvy21(test);
            Assert.AreEqual(expected.Northing, actual.Northing, ToleranceSvy21);
            Assert.AreEqual(expected.Easting, actual.Easting, ToleranceSvy21);
        }
    }
}

File: C++/tests/svy21test.cpp
#include "../svy21.h"
#include "gtest/gtest.h"

static const double ToleranceLatLon = 1e-10;
static const double ToleranceSvy21 = 1e-3;

TEST(Svy21Test, Svy21ToLatLon) {
	SVY21 svy21;
	double latlon[2];

	// SVY21 reference point.
	svy21.SVY21ToLatLon(38744.572, 28001.642, &latlon[0], &latlon[1]);
	ASSERT_NEAR(1.366666, latlon[0], ToleranceLatLon);
	ASSERT_NEAR(103.833333, latlon[1], ToleranceLatLon);

	// Corner of Ang Mo Kio 66kV Substation.
	svy21.SVY21ToLatLon(39105.269, 30629.967, &latlon[0], &latlon[1]);
	ASSERT_NEAR(1.3699278977737488, latlon[0], ToleranceLatLon);
	ASSERT_NEAR(103.856950349764668, latlon[1], ToleranceLatLon);

	// Corner of Jurong Lake Canal 400kV Cable Bridge.
	svy21.SVY21ToLatLon(36307.704, 16272.970, &latlon[0], &latlon[1]);
	ASSERT_NEAR(1.3446255443241177, latlon[0], ToleranceLatLon);
	ASSERT_NEAR(103.72794378041792, latlon[1], ToleranceLatLon);

	// Corner of Sembawang 66kV Substation.
	svy21.SVY21ToLatLon(48187.789, 27720.130, &latlon[0], &latlon[1]);
	ASSERT_NEAR(1.4520670518379692, latlon[0], ToleranceLatLon);
	ASSERT_NEAR(103.83080332777138, latlon[1], ToleranceLatLon);
}

TEST(Svy21Test, LatLonToSvy21) {
	SVY21 svy21;
	double northeast[2];

	// SVY21 Reference Point.
	svy21.latLonToSVY21(1.366666, 103.833333, &northeast[0], &northeast[1]);
	ASSERT_NEAR(38744.572, northeast[0], ToleranceSvy21);
	ASSERT_NEAR(28001.642, northeast[1], ToleranceSvy21);

	// Corner of Ang Mo Kio 66kV Substation.
	svy21.latLonToSVY21(1.3699278977737488, 103.85695034976466, &northeast[0], &northeast[1]);
	ASSERT_NEAR(39105.269, northeast[0], ToleranceSvy21);
	ASSERT_NEAR(30629.967, northeast[1], ToleranceSvy21);

	// Corner of Jurong Lake Canal 400kV Cable Bridge.
	svy21.latLonToSVY21(1.3446255443241177, 103.72794378041792, &northeast[0], &northeast[1]);
	ASSERT_NEAR(36307.704, northeast[0], ToleranceSvy21);
	ASSERT_NEAR(16272.970, northeast[1], ToleranceSvy21);

	// Corner of Sembawang 66kV Substation.
	svy21.latLonToSVY21(1.4520670518379692, 103.83080332777138, &northeast[0], &northeast[1]);
	ASSERT_NEAR(48187.789, northeast[0], ToleranceSvy21);
	ASSERT_NEAR(27720.130, northeast[1], ToleranceSvy21);
}

int main(int argc, char **argv) {
	::testing::InitGoogleTest(&argc, argv);
	return RUN_ALL_TESTS();
}
